torch/cuda/_sanitizer.py

```diff
@@ -291,13 +292,43 @@ def record_state(self, event: EventId, stream: StreamId) -> None:
         self._ensure_stream_exists(stream)
         self.recorded_sync_states[event] = self.current_sync_states[stream].copy()
 
-    def state_wait_for_event(self, stream: StreamId, event: EventId) -> None:
+    def _state_wait_for_other(
+        self, state: Dict[StreamId, SeqNum], other: Dict[StreamId, SeqNum]
+    ) -> None:
+        for stream, seq_num in other.items():
+            state[stream] = max(state.get(stream, -1), seq_num)
+
+    def stream_wait_for_event(self, stream: StreamId, event: EventId) -> None:
+        self._ensure_stream_exists(stream)
+        self._ensure_event_exists(event)
+        self._state_wait_for_other(
+            self.current_sync_states[stream], self.recorded_sync_states[event]
+        )
+
+    def all_streams_wait_for_event(self, event: EventId) -> None:
         self._ensure_event_exists(event)
+        for stream in self.current_sync_states.keys():
+            self.stream_wait_for_event(stream, event)
+
+        self._state_wait_for_other(
+            self.host_sync_state, self.recorded_sync_states[event]
+        )
+
+    def all_streams_wait_for_stream(self, stream: StreamId) -> None:
         self._ensure_stream_exists(stream)
-        for other_stream, seq_num in self.recorded_sync_states[event].items():
-            self.current_sync_states[stream][other_stream] = max(
-                self.current_sync_states[stream].get(other_stream, -1), seq_num
-            )
+        for state in self.current_sync_states.values():
+            self._state_wait_for_other(state, self.current_sync_states[stream])
+
+        self._state_wait_for_other(
+            self.host_sync_state, self.current_sync_states[stream]
+        )
+
+    def all_streams_sync_to_seq_num(self, seq_num: SeqNum) -> None:
+        for stream in self.current_sync_states.keys():
+            self.host_sync_state[stream] = seq_num
```

Comment:
I want to point out a minor but interesting thing. It's true that the sequence number is a single increasing value among all streams, however until now we always used it in a way that was "namespaced" by stream. What I mean is that we kind of never used a sequence number in isolation, we always paired it with a certain stream. This means that in practice we could have replaced the one "global" sequence with many individual per-stream sequences of numbers and everything would have kept working in exactly the same way.

This change now for the first time breaks this equivalence. However, it doesn't have to. I believe an equivalent way of writing these lines would be:
```suggestion
        for stream, state in self.current_sync_states.items():
            self.host_sync_state[stream] = state[stream]
```
which means that we update the host state so that, for each stream, it comes after the last kernel launched on that stream.

I'm not asking for this code to be changed, I just want to point out that there are two ways of thinking about this, and I can't necessarily see why one would be better than the other, I just wanted you to be aware.

Reply:
I like your version more; mine broke the implicit rule that each stream was "in possession" of its synchronization sequential number.